### PR TITLE
Core service

### PR DESCRIPTION
**Pull request by borispovod**
Still draft, as still have to implement block producers function and improve tests.

### CODE REVIEW

File: core/chain.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
+/**
+ * @module core/chain
+ */
+'use strict';
+
+const chainData = require('core/chaindata');
+
+/**
+ * Returns balance of account.
```
Comment (romfrolov) [inline core/chain.js]: For return value there is **@return** (or **@returns** tag) in jsdoc. It's a good practice to describe what function does in a more generic way.

File: core/chain.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+ * @returns {Promise.<Number>} Balance of account.
+ */
+exports.getBalance = async function getBalance(address) {
+    const account = await getAccountFromLatestBlock(address);
+
+    return account.balance;
+};
+
+/**
+ * Returns stake of account.
```
Comment (romfrolov) [inline core/chain.js]: See [comment above](https://github.com/WingsDao/griffin-consensus-poc/pull/30/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json#r273210628).

File: core/chain.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    if (!account.certificates.length) {
+        return 0;
+    }
+
+    return account.locked;
+};
+
+/**
+ * Returns votes for account.
```
Comment (romfrolov) [inline core/chain.js]: See [comment above](https://github.com/WingsDao/griffin-consensus-poc/pull/30/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json#r273210628).

File: core/chain.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        }
+
+        return rawDelegates;
+    }, {});
+
+    return rawDelegates;
+};
+
+/**
+ * Returns account by address from latetst block.
```
Comment (romfrolov) [inline core/chain.js]: See [comment above](https://github.com/WingsDao/griffin-consensus-poc/pull/30/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json#r273210628).

File: genesis.json [hunk 5]
```diff
@@ -16,5 +16,6 @@
         {"0x0000000000000000000000000000000000000000": {"balance": 0}},
         {"0x0000000000000000000000000000000000000000": {"balance": 0}},
         {"0x0000000000000000000000000000000000000000": {"balance": 0}}
-    ]
+    ],
+    "state": []
```
Comment (romfrolov) [inline genesis.json:11]: Please, check [genesis library](https://github.com/WingsDao/griffin-consensus-poc/blob/delegates/lib/genesis.js#L31-L38) for compatibility with this change.

File: test/unit/chain.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        account = Account(SECRET_KEYS[0]);
+        delegate = Account(SECRET_KEYS[1]);
+    });
+
+    it('get account balance', async () => {
+        const balance = await chain.getBalance(account.getStrAddress());
+
+        balance.should.be.a('number');
+
+        console.log('Balance: ', balance);
```
Comment (romfrolov) [inline test/unit/chain.js]: If you are passing arguments using comma separation extra space will be added automatically.
```suggestion
        console.log('Balance:', balance);
```

File: test/unit/chain.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    });
+
+    xit('get account stake', async () => {
+        const stake = await chain.getStake(account.getStrAddress());
+
+        stake.should.be.a('number');
+
+        console.log('Stake: ', stake);
```
Comment (romfrolov) [inline test/unit/chain.js]:
```suggestion
        console.log('Stake:', stake);
```

File: test/unit/chain.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    });
+
+    it('get account votes', async () => {
+        const votes = await chain.getVotesFor(delegate.getStrAddress());
+
+        console.log('Votes: ', votes);
```
Comment (romfrolov) [inline test/unit/chain.js]:
```suggestion
        console.log('Votes:', votes);
```

File: test/unit/chain.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    });
+
+    it('get all delegates', async () => {
+        delegates = await chain.getDelegates();
+
+        delegates.should.be.a('array');
+        console.log('Delegates: ', delegates);
```
Comment (romfrolov) [inline test/unit/chain.js]:
```suggestion
        console.log('Delegates:', delegates);
```

File: test/unit/chain.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    });
+
+    it('get active delegates', async () => {
+        const activeDelegates = await chain.getActiveDelegates();
+
+        console.log('Active delegates: ', activeDelegates);
```
Comment (romfrolov) [inline test/unit/chain.js]:
```suggestion
        console.log('Active delegates:', activeDelegates);
```

File: test/unit/chain.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    });
+
+    it('get successor delegates', async () => {
+        const successorDelegates = await chain.getSuccessorDelegates();
+
+        console.log('Successor delegates', successorDelegates);
```
Comment (romfrolov) [inline test/unit/chain.js]:
```suggestion
        console.log('Successor delegates:', successorDelegates);
```

File: test/unit/chain.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    });
+
+    it('get delegatets count', async () => {
```
Comment (romfrolov) [inline test/unit/chain.js]:
```suggestion
    it('get delegates count', async () => {
```

File: test/unit/chain.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
+/**
+ * @module test/unit/chain
+ */
+
+'use strict';
+
+require('chai').should();
+
+const Account = require('core/account');
+const chain = require('core/chain');
```
Comment (romfrolov) [inline test/unit/chain.js]:
```suggestion
const chain   = require('core/chain');
```

File: test/unit/chain.js [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+const SECRET_KEYS = [
+    Buffer.from('557dce58018cf502a32b9b7723024805399350d006a4f71c3b9f489f7085cb50', 'hex'),
```
Comment (romfrolov) [inline test/unit/chain.js]: What is the reason for hard-coded secret keys? If no secret key is provided, random secret key will be generated on each account creation (see [core/account.js](https://github.com/WingsDao/griffin-consensus-poc/blob/delegates/core/account.js#L34-L36)).

**Review by romfrolov — DISMISSED** [review]
Please see comments to resolve issues.

File: core/account.js [hunk 15]
```diff
@@ -87,6 +87,15 @@ Account.prototype.tx = function tx(to, value, data='0x00') {
     return '0x' + tx.serialize().toString('hex');
 };
 
+/**
+ * Get address of account as hex string.
+ *
+ * @returns {string} Address as hex string.
+ */
+Account.prototype.getStrAddress = function getStrAddress() {
```
Comment (damirka) [inline core/account.js]: It's probably should be called 'getAddressHex' or 'getHexAddress'. Buffer can be transformed to string in a numerous ways, so calling it 'str' is not the best way of describing what this function does.

File: test/unit/chain.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+    it('check account is delegate', async () => {
+        const isDelegate = await chain.isDelegate(delegates[0]);
+
+        isDelegate.should.be.true;
+    });
+
+
+
+});
```
Comment (damirka) [inline test/unit/chain.js]: New line in the end of file is obligatory.

**Review by damirka — DISMISSED** [review]

**Comment by borispovod** [comment]
Need to clarify difference between blockchain.js functional and chain.js functional and probably merge them

File: core/blockchain.js [hunk 17]
```diff
@@ -55,6 +56,152 @@ exports.handleTransaction = function handleTransaction(tx, block) {
...
+
+/**
+ * Returns 31 active delegates.
+ *
+ * @returns {Promise.<Object[]>}
+ */
+exports.getActiveDelegates = async function getActiveDelegates() {
+    const delegates = await this.getDelegates();
+
+    return delegates.slice(0, 31);
```
Comment (damirka) [inline core/blockchain.js]: 31 is a magic number.
It has to be defined as module-level constant and exported outside.
i.e.
```JavaScript
const ACTIVE_DELEGATES_COUNT = exports.ACTIVE_DELEGATES_COUNT = 31;
```

File: core/blockchain.js [hunk 18]
```diff
@@ -55,6 +56,152 @@ exports.handleTransaction = function handleTransaction(tx, block) {
...
+    return delegates;
+};
+
+/**
+ * Returns 31 active delegates.
+ *
+ * @returns {Promise.<Object[]>}
+ */
+exports.getActiveDelegates = async function getActiveDelegates() {
+    const delegates = await this.getDelegates();
```
Comment (damirka) [inline core/blockchain.js]: It's better not to use 'this' context with static methods. 
You can define `getDelegates` as a function, export it and use in methods like this.

File: core/blockchain.js [hunk 19]
```diff
@@ -186,3 +333,61 @@ function getBlockProducer(block, finalRandomNumber) {
...
+ */
+async function getAccount(address, block) {
+    const account = block.state.find(account => account.address == address);
+
+    if (!account) {
+        throw `Account '0x${address.toString('hex')}' not found`;
+    }
+
+    return account;
+}
```
Comment (damirka) [inline core/blockchain.js]: Empty line required.

File: core/account.js [hunk 20]
```diff
@@ -79,6 +79,15 @@ Account.prototype.tx = function tx(to, value, data='0x00') {
     return '0x' + tx.serialize().toString('hex');
 };
 
+/**
+ * Get address of account as hex string.
+ *
+ * @returns {string} Address as hex string.
+ */
+Account.prototype.getHexAddress = function getHexAddress() {
```
Comment (damirka) [inline core/account.js:87]: 👍🏻👍🏻👍🏻

**Review by damirka — CHANGES_REQUESTED** [review]

**Review by damirka — APPROVED** [review]

**Pull request merged**